### PR TITLE
feat(trainer): emergency checkpointing on crashes & SIGTERM/SIGINT

### LINKED ISSUE #38961: 🛠️ More robust Trainer
### Feature request

Introduce a **failure-safe training mode** in the `Trainer` API (e.g., via a flag like `enable_emergency_checkpoint=True`) that ensures training progress is saved **automatically upon unexpected failure**, such as hardware interruptions, out-of-memory errors, or bugs during training. This allows users to resume from the **last known safe state**, even if a checkpoint hasn't yet been saved via the normal `save_strategy`.

### Motivation

Current training resumption methods rely on:

- Periodic checkpoints using `save_steps` or `save_strategy`
- Manual try/except logic for error handling and saving state

However, these methods have limitations:

- Saving too frequently slows down training due to I/O overhead
- Saving only at the end of an epoch (`save_strategy="epoch"`) risks **losing progress if failure occurs mid-epoch**
- Manual error handling is fragile and inconsistent between users

This can lead to **wasted compute time** and inability to resume training effectively.

### Your contribution

N/A

### PR DESCRIPTION
**Pull request by AyushSharma173**
# What does this PR do?
Adds *failure-safe training* to the 🤗 `Trainer`.

* **New flag** `enable_emergency_checkpoint` in `TrainingArguments`
  * default: `False` (opt-in)
* **Automatic emergency save** when training ends unexpectedly
  * un-handled exception inside `Trainer.train`
  * external `SIGTERM`/`SIGINT` (e.g. job pre-emption, Ctrl-C)
* Checkpoint is written to  
  `<output_dir>/checkpoint-emergency`
  and contains model weights, optimizer, scheduler, scaler, RNG state
  plus a minimal `trainer_state.json`.
* Resume works transparently via  
  `TrainingArguments(..., resume_from_checkpoint=".../checkpoint-emergency")`.

### Motivation

Issue #38961 requests a robust way to avoid losing progress when a run is killed mid-epoch or by OOM/infra errors.  
Current work-arounds (very frequent `save_steps` or a user-level `try/except`) are slow or brittle.  
This PR makes the feature one-line (`enable_emergency_checkpoint=True`) and cost-free when disabled.

### Implementation details

* **`training_args.py`** – new dataclass field `enable_emergency_checkpoint`.
* **`trainer.py`**
  * registers signal & `atexit` handlers when the flag is on
  * wraps `train()` loop in `try/except` and calls `_common_emergency_save`
  * `_common_emergency_save` is idempotent & rank-safe (`_emergency_save_running / _completed` flags).
* **Tests**
  * `tests/trainer/test_emergency_ckpt.py`
    * verifies flag round-trip
    * asserts emergency folder is created on crash and training can resume
    * checks that opting-out leaves no folder

Fixes # (issue)
Closes #38961.

### Before submitting

- [x] I read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request).
- [x] The change is discussed in #38961.
- [x] Added unit tests.
- [x] Ran `make fixup && make quality` locally.
- [ ] Documentation not updated (trainer argument + short snippet) – **TODO**

## Who can review?
Trainer reviewers: @zach-huggingface, @SunMarc, @qgallouedec

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2190,25 +2221,33 @@ def train(
...
             self._inner_training_loop, self._train_batch_size, args.auto_find_batch_size
         )
-        if args.push_to_hub:
-            try:
-                # Disable progress bars when uploading models during checkpoints to avoid polluting stdout
-                hf_hub_utils.disable_progress_bars()
+
+        try:
+            if args.push_to_hub:
+                try:
```
Comment (qgallouedec) [inline src/transformers/trainer.py]: Wrapping the whole training loop into a try/expect would make the use of debugger quite challenging, no?

Comment (AyushSharma173) [reply src/transformers/trainer.py]: Good point. I’ve adjusted the implementation so the try except now only surrounds the call to inner_training_loop (via a small _run_inner_training_loop_with_emergency_ckpt helper).
Everything that happens earlier in train is executed outside a try so debuggers will still break exactly at the offending line. The emergency checkpoint logic is therefore triggered only for failures that occur after training has actually started, which is where we have state worth saving.

**Comment by ArthurZucker** [comment]
cc @SunMarc as well

File: src/transformers/trainer.py (hunk 1 above)
Comment (qgallouedec) [reply src/transformers/trainer.py]: Thanks for narrowing the scope of the try/epxect, although I think that the remark remains for everything inside the training loop.
Not saying that's it's a major issue for this feature, but I feel like it could make the development work a lot more annoying

**Comment by yaswanth19** [comment]
CC: @SunMarc for your attn here

**Comment by yaswanth19** [comment]
hey @AyushSharma173 I don't think saving it in `<output_dir>/checkpoint-emergency`  folder makes sense. When we retrigger the training run then it should pickup the last saved checkpoint/emergency checkpoint from the output dir itself. So the checkpoint should be saved in output dir with the last step number before it crashed like `output_dir/checkpoint-123`

**Comment by yaswanth19** [comment]
@SunMarc A gentle ping here for review. Also @AyushSharma173 LMK if you don't have the bandwidth to continue as I have some time now and can it pick up.

**Comment by AyushSharma173** [comment]
@yaswanth19 Hey sure. You can pickup this issue and try to address the concerns raised so far. I dont have time for next week, so feel free.

**Comment by efazal** [comment]
Hi @AyushSharma173, This is great work! I've been working on a similar feature and only now came across your PR.
We wanted to solve a problem common faced by many people on a shared cluster, where their training jobs get preempted by high priority jobs and they lose progress, or for any reasons users would like to pause their training jobs and be able to resume without loosing progress up until the point of resume.

I took a proactive approach as opposed to your reactive approach to solve this and named it Just-in-Time (JIT) checkpointing. And I also wanted to ensure the trainer code was not disrupted a lot. For that reason I took a modular approach. I'm also leveraging SIGTERM signals for this implementation, and the key difference would we that I'm doing the checkpoint asynchronously. Please have a look at my approach implemented in my fork and let me know your thoughts as I would love to contribute these changes as part of this PR or even a fresh PR if you think that would be wise:
https://github.com/huggingface/transformers/compare/main...efazal:transformers:main

cc: @qgallouedec @yaswanth19

**Comment by qgallouedec** [comment]
what about just:

```python
try:
    trainer.train()
except:
    trainer.save_model()
```

**Comment by efazal** [comment]
Hi @AyushSharma173 @qgallouedec @SunMarc, I was wondering if I could pickup this issue and continue working on the feature. In my previous comment I've added a link to my [fork](https://github.com/huggingface/transformers/compare/main...efazal:transformers:main) with the approach I took to implement emergency checkpoint. It addresses all the review comments here so far. Please have a look and let me know if I can go ahead and submit a new revision to this PR. Thank you.
